cpp/references/data_structures/list_digraph.cpp

```diff
@@ -3,636 +3,570 @@
-	LemonArc arc = LemonDigraph::arcFromId(p_arc);
-	ERR_FAIL_COND(!_graph.valid(arc));
-	_graph.erase(arc);
+void ListDigraph::_erase_arc(Variant p_arc) {
+	ERR_FAIL_COND(p_arc.get_type() != Variant::OBJECT);
+	erase_arc(p_arc);
 }
 
-void ListDigraph::erase_arcs(Vector<ArcID> p_arcs) {
+void ListDigraph::erase_arcs(const Array &p_arcs) {
 	for (int i = 0; i < p_arcs.size(); i++) {
-		ArcID id = p_arcs[i];
-		LemonArc arc = LemonDigraph::arcFromId(id);
-		ERR_CONTINUE(!_graph.valid(arc));
-		_graph.erase(arc);
-	}
-}
+		Ref<Arc> arc = p_arcs[i];
+		ERR_CONTINUE(arc.is_null());
 
-void ListDigraph::reserve_arcs(int p_amount) {
-	_graph.reserveArc(p_amount);
-}
+		LemonDigraph::Arc handle = arc->get_handle();
+		ERR_CONTINUE(!_graph.valid(handle));
 
-Vector<ArcID> ListDigraph::get_arcs() {
-	Vector<ArcID> ret;
-	for (LemonArcIt it(_graph); it != lemon::INVALID; ++it) {
-		ret.push_back(LemonDigraph::id(it));
+		_graph.erase(handle);
 	}
-	return ret;
 }
-
-void ListDigraph::clear() {
-	_graph.clear();
+void ListDigraph::_erase_arcs(Variant p_arcs) {
+	ERR_FAIL_COND(p_arcs.get_type() != Variant::ARRAY);
+	erase_arcs(p_arcs);
 }
 
-void ListDigraph::set_data(ListDigraph::GraphComponent p_type, int64_t p_id, Variant p_data) {
-	switch (p_type)
-	{
-	case ListDigraph::VERTEX:
-	{
-		LemonNode node = LemonDigraph::nodeFromId(p_id);
-		ERR_FAIL_COND(!_graph.valid(node));
-		_vertex_map.set(node, p_data);
-		return;
-	}
-	case ListDigraph::ARC:
-	{
-		LemonArc arc = LemonDigraph::arcFromId(p_id);
-		ERR_FAIL_COND(!_graph.valid(arc));
-		_arc_map.set(arc, p_data);
-		return;
-	}
-	default:
-	{
-		ERR_FAIL();
-		break;
-	}
-	}
+void ListDigraph::_reserve_arcs(Variant p_amount) {
+	ERR_FAIL_COND(p_amount.get_type() != Variant::INT);
+	reserve_arcs(p_amount);
 }
 
-Variant ListDigraph::get_data(ListDigraph::GraphComponent p_type, int64_t p_id) {
-	switch (p_type)
-	{
-	case ListDigraph::VERTEX:
-	{
-		LemonNode node = LemonDigraph::nodeFromId(p_id);
-		ERR_FAIL_COND_V(!_graph.valid(node), Variant());
-		return _vertex_map.operator[](node);
-	}
-	case ListDigraph::ARC:
-	{
-		LemonArc arc = LemonDigraph::arcFromId(p_id);
-		ERR_FAIL_COND_V(!_graph.valid(arc), Variant());
-		return _arc_map.operator[](arc);
-	}
-	default:
-	{
-		ERR_FAIL();
-		break;
-	}
+Array ListDigraph::_get_arcs() {
+	Array ret;
+	for (LemonDigraph::ArcIt it(_graph); it != lemon::INVALID; ++it) {
+		ret.append(get_instance(it));
 	}
+	return ret;
 }
 
-void ListDigraph::update_data(ListDigraph::GraphComponent p_type, const HashMap<int64_t, Variant> &p_data) {
-	HashMap<int64_t, Variant>::const_iterator an_id_to_value;
-	switch (p_type)
-	{
-	case ListDigraph::VERTEX:
-		for (an_id_to_value = p_data.begin(); an_id_to_value != p_data.end(); ++an_id_to_value) {
-			
-			LemonNode node = LemonDigraph::nodeFromId(an_id_to_value->first);
-			ERR_CONTINUE(!_graph.valid(node));
-			_vertex_map.set(node, an_id_to_value->second);
+void ListDigraph::update_data(const Dictionary &p_data) {
```

Comment:
Since we have the ability to, I think we should adjust the implementation of these batch methods to handle either passing in Object keys OR ID keys, that way, if users don't have the objects, but still have the unique ID, they can just as easily use those instead (or even use both at the same time, for whatever reason). Both inputs act as a primary key to the data, so I think we should offer that. What do you think?

Reply:
yeah.... but passing the object let us know if it is a vertex or an arc.. passing an id doesn't let us distinguish... we can make special methods for those... `update_data_by_id()`

Reply:
Ah, I see what you mean. Yeah, that makes sense.